{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allows for multiple cell outputs\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "#importing necessary libraries\n",
    "import os\n",
    "import re\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import glob\n",
    "import gzip\n",
    "from lxml import etree\n",
    "import time\n",
    "import nltk\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import gzip\n",
    "import csv\n",
    "import time\n",
    "from lxml import etree\n",
    "\n",
    "journals = ['Antonie van Leeuwenhoek', 'International journal of systematic and evolutionary microbiology', 'Biochemical and biophysical research communications']\n",
    "\n",
    "path = 'TIMC-quinoneMining/pubmed-data'\n",
    "xml_files = glob.glob(os.path.join(path, '*.xml.gz'))\n",
    "\n",
    "file_limit = 1  # set number of files to process\n",
    "xml_files = xml_files[:file_limit]  # apply file limit\n",
    "\n",
    "start_time = time.time()  # start timing\n",
    "\n",
    "count = 0  # initialize counter\n",
    "\n",
    "with open('output.csv', 'w', newline='', encoding='utf-8') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['Journal Name', 'Date of Publication', 'PubMedID', 'Title', 'Abstract'])\n",
    "\n",
    "    for i, file_path in enumerate(xml_files):\n",
    "        try:\n",
    "            with gzip.open(file_path, 'rb') as f:\n",
    "                content = f.read()\n",
    "        except OSError as e:\n",
    "            print(f\"Error opening file: {file_path}. Reason: {str(e)}\")\n",
    "            continue\n",
    "\n",
    "        try:\n",
    "            root = etree.fromstring(content)\n",
    "        except etree.XMLSyntaxError as e:\n",
    "            print(f\"Error parsing XML: {file_path}. Reason: {str(e)}\")\n",
    "            continue\n",
    "\n",
    "        for article in root.findall('.//PubmedArticle'):\n",
    "            abstract_element = article.find('.//AbstractText')\n",
    "            journal_name_element = article.find('.//Journal/Title')\n",
    "\n",
    "            if abstract_element is None or journal_name_element is None or journal_name_element.text not in journals:\n",
    "                continue\n",
    "\n",
    "            title = article.findtext('.//ArticleTitle', default='')\n",
    "            abstract = abstract_element.text or ''\n",
    "            pubmed_id = article.findtext('.//PMID', default='')\n",
    "\n",
    "            if not pubmed_id.isnumeric():\n",
    "                print(f\"Unexpected PubMedID in file: {file_path}. Skipping this record.\")\n",
    "                continue\n",
    "\n",
    "            pub_date_element = article.find('.//PubDate')\n",
    "            if pub_date_element is not None:\n",
    "                pub_year = pub_date_element.findtext('.//Year', default='')\n",
    "                pub_month = pub_date_element.findtext('.//Month', default='')\n",
    "                pub_date = f\"{pub_year}-{pub_month}\"\n",
    "            else:\n",
    "                pub_date = ''\n",
    "\n",
    "            writer.writerow([journal_name_element.text, pub_date, pubmed_id, title, abstract])\n",
    "            count += 1  # increment counter\n",
    "\n",
    "end_time = time.time()  # end timing\n",
    "\n",
    "print(f\"Process took {end_time - start_time} seconds.\")\n",
    "print(f\"Total articles processed from specified journals: {count}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 79 entries, 0 to 78\n",
      "Data columns (total 5 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   Journal Name         79 non-null     object\n",
      " 1   Date of Publication  79 non-null     object\n",
      " 2   PubMedID             79 non-null     int64 \n",
      " 3   Title                69 non-null     object\n",
      " 4   Abstract             79 non-null     object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 3.2+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Journal Name</th>\n",
       "      <th>Date of Publication</th>\n",
       "      <th>PubMedID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145673</td>\n",
       "      <td>Kineobactrum sediminis gen. nov., sp. nov., is...</td>\n",
       "      <td>A novel Gram-stain-negative, rod-shaped marine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145674</td>\n",
       "      <td>Cutaneotrichosporon suis sp. nov., a lipolytic...</td>\n",
       "      <td>Two conspecific yeast strains, which based on ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145675</td>\n",
       "      <td>Shewanella psychromarinicola sp. nov., a psych...</td>\n",
       "      <td>Two Gram-stain-negative, rod-shaped, facultati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Oct</td>\n",
       "      <td>31145676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A rod-shaped, spore-forming, thermophilic, che...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Oct</td>\n",
       "      <td>31145678</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Two Gram-stain-negative, catalase- and oxidase...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Journal Name Date of Publication  \\\n",
       "0  International journal of systematic and evolut...            2019-Aug   \n",
       "1  International journal of systematic and evolut...            2019-Aug   \n",
       "2  International journal of systematic and evolut...            2019-Aug   \n",
       "3  International journal of systematic and evolut...            2019-Oct   \n",
       "4  International journal of systematic and evolut...            2019-Oct   \n",
       "\n",
       "   PubMedID                                              Title  \\\n",
       "0  31145673  Kineobactrum sediminis gen. nov., sp. nov., is...   \n",
       "1  31145674  Cutaneotrichosporon suis sp. nov., a lipolytic...   \n",
       "2  31145675  Shewanella psychromarinicola sp. nov., a psych...   \n",
       "3  31145676                                                NaN   \n",
       "4  31145678                                                NaN   \n",
       "\n",
       "                                            Abstract  \n",
       "0  A novel Gram-stain-negative, rod-shaped marine...  \n",
       "1  Two conspecific yeast strains, which based on ...  \n",
       "2  Two Gram-stain-negative, rod-shaped, facultati...  \n",
       "3  A rod-shaped, spore-forming, thermophilic, che...  \n",
       "4  Two Gram-stain-negative, catalase- and oxidase...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Journal Name</th>\n",
       "      <th>Date of Publication</th>\n",
       "      <th>PubMedID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2019-07</td>\n",
       "      <td>31171360</td>\n",
       "      <td>Crystal structure of pentameric shell protein ...</td>\n",
       "      <td>Carboxysome, encapsulating an enzymatic core w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2019-07</td>\n",
       "      <td>31171361</td>\n",
       "      <td>miR-374a/Myc axis modulates iron overload-indu...</td>\n",
       "      <td>The transformation of hepatic stellate cells (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Biochemical and biophysical research communica...</td>\n",
       "      <td>2019-07</td>\n",
       "      <td>31171362</td>\n",
       "      <td>In situ reactivity of electrochemically genera...</td>\n",
       "      <td>There is enough proof to believe that free-rad...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Antonie van Leeuwenhoek</td>\n",
       "      <td>2019-Oct</td>\n",
       "      <td>31172329</td>\n",
       "      <td>Flavisolibacter galbus sp. nov., isolated from...</td>\n",
       "      <td>A Gram-stain negative, non-motile, and yellow-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Antonie van Leeuwenhoek</td>\n",
       "      <td>2019-Nov</td>\n",
       "      <td>31172330</td>\n",
       "      <td>Vibrio profundi sp. nov., isolated from a deep...</td>\n",
       "      <td>A Gram-stain negative, rod-shaped, facultative...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Journal Name Date of Publication  \\\n",
       "74  Biochemical and biophysical research communica...             2019-07   \n",
       "75  Biochemical and biophysical research communica...             2019-07   \n",
       "76  Biochemical and biophysical research communica...             2019-07   \n",
       "77                            Antonie van Leeuwenhoek            2019-Oct   \n",
       "78                            Antonie van Leeuwenhoek            2019-Nov   \n",
       "\n",
       "    PubMedID                                              Title  \\\n",
       "74  31171360  Crystal structure of pentameric shell protein ...   \n",
       "75  31171361  miR-374a/Myc axis modulates iron overload-indu...   \n",
       "76  31171362  In situ reactivity of electrochemically genera...   \n",
       "77  31172329  Flavisolibacter galbus sp. nov., isolated from...   \n",
       "78  31172330  Vibrio profundi sp. nov., isolated from a deep...   \n",
       "\n",
       "                                             Abstract  \n",
       "74  Carboxysome, encapsulating an enzymatic core w...  \n",
       "75  The transformation of hepatic stellate cells (...  \n",
       "76  There is enough proof to believe that free-rad...  \n",
       "77  A Gram-stain negative, non-motile, and yellow-...  \n",
       "78  A Gram-stain negative, rod-shaped, facultative...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Building Quinone Dictionary\n",
    "df = pd.read_csv(\"output.csv\")\n",
    "df.info()\n",
    "df.head()\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 420 entries, 0 to 419\n",
      "Data columns (total 5 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   Journal Name         420 non-null    object\n",
      " 1   Date of Publication  420 non-null    object\n",
      " 2   PubMedID             420 non-null    int64 \n",
      " 3   Title                346 non-null    object\n",
      " 4   Abstract             419 non-null    object\n",
      "dtypes: int64(1), object(4)\n",
      "memory usage: 16.5+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Journal Name</th>\n",
       "      <th>Date of Publication</th>\n",
       "      <th>PubMedID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145673</td>\n",
       "      <td>Kineobactrum sediminis gen. nov., sp. nov., is...</td>\n",
       "      <td>A novel Gram-stain-negative, rod-shaped marine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145674</td>\n",
       "      <td>Cutaneotrichosporon suis sp. nov., a lipolytic...</td>\n",
       "      <td>Two conspecific yeast strains, which based on ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Aug</td>\n",
       "      <td>31145675</td>\n",
       "      <td>Shewanella psychromarinicola sp. nov., a psych...</td>\n",
       "      <td>Two Gram-stain-negative, rod-shaped, facultati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Oct</td>\n",
       "      <td>31145676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A rod-shaped, spore-forming, thermophilic, che...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2019-Oct</td>\n",
       "      <td>31145678</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Two Gram-stain-negative, catalase- and oxidase...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Journal Name Date of Publication  \\\n",
       "0  International journal of systematic and evolut...            2019-Aug   \n",
       "1  International journal of systematic and evolut...            2019-Aug   \n",
       "2  International journal of systematic and evolut...            2019-Aug   \n",
       "3  International journal of systematic and evolut...            2019-Oct   \n",
       "4  International journal of systematic and evolut...            2019-Oct   \n",
       "\n",
       "   PubMedID                                              Title  \\\n",
       "0  31145673  Kineobactrum sediminis gen. nov., sp. nov., is...   \n",
       "1  31145674  Cutaneotrichosporon suis sp. nov., a lipolytic...   \n",
       "2  31145675  Shewanella psychromarinicola sp. nov., a psych...   \n",
       "3  31145676                                                NaN   \n",
       "4  31145678                                                NaN   \n",
       "\n",
       "                                            Abstract  \n",
       "0  A novel Gram-stain-negative, rod-shaped marine...  \n",
       "1  Two conspecific yeast strains, which based on ...  \n",
       "2  Two Gram-stain-negative, rod-shaped, facultati...  \n",
       "3  A rod-shaped, spore-forming, thermophilic, che...  \n",
       "4  Two Gram-stain-negative, catalase- and oxidase...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Journal Name</th>\n",
       "      <th>Date of Publication</th>\n",
       "      <th>PubMedID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>415</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2016-Dec</td>\n",
       "      <td>27618795</td>\n",
       "      <td>Labrenzia salina sp. nov., isolated from the r...</td>\n",
       "      <td>A novel, halophilic, motile, rod-shaped, Gram-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2016-Dec</td>\n",
       "      <td>27619232</td>\n",
       "      <td>Isolation and characterization of a novel Gram...</td>\n",
       "      <td>The taxonomic position of a Gram-stain negativ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2016-Dec</td>\n",
       "      <td>27620694</td>\n",
       "      <td>Patulibacter brassicae sp. nov., isolated from...</td>\n",
       "      <td>A novel actinobacterial strain, designated SDT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2016-Dec</td>\n",
       "      <td>27620848</td>\n",
       "      <td>Genome-based phylogeny and taxonomy of the 'En...</td>\n",
       "      <td>Understanding of the phylogeny and interrelati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>International journal of systematic and evolut...</td>\n",
       "      <td>2016-Dec</td>\n",
       "      <td>27620889</td>\n",
       "      <td>Saccharomonospora xiaoerkulensis sp. nov., iso...</td>\n",
       "      <td>A novel actinomycete, strain TRM 41495T, was i...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Journal Name Date of Publication  \\\n",
       "415  International journal of systematic and evolut...            2016-Dec   \n",
       "416  International journal of systematic and evolut...            2016-Dec   \n",
       "417  International journal of systematic and evolut...            2016-Dec   \n",
       "418  International journal of systematic and evolut...            2016-Dec   \n",
       "419  International journal of systematic and evolut...            2016-Dec   \n",
       "\n",
       "     PubMedID                                              Title  \\\n",
       "415  27618795  Labrenzia salina sp. nov., isolated from the r...   \n",
       "416  27619232  Isolation and characterization of a novel Gram...   \n",
       "417  27620694  Patulibacter brassicae sp. nov., isolated from...   \n",
       "418  27620848  Genome-based phylogeny and taxonomy of the 'En...   \n",
       "419  27620889  Saccharomonospora xiaoerkulensis sp. nov., iso...   \n",
       "\n",
       "                                              Abstract  \n",
       "415  A novel, halophilic, motile, rod-shaped, Gram-...  \n",
       "416  The taxonomic position of a Gram-stain negativ...  \n",
       "417  A novel actinobacterial strain, designated SDT...  \n",
       "418  Understanding of the phylogeny and interrelati...  \n",
       "419  A novel actinomycete, strain TRM 41495T, was i...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 108 entries, 40 to 419\n",
      "Data columns (total 8 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   Journal Name         108 non-null    object\n",
      " 1   Date of Publication  108 non-null    object\n",
      " 2   PubMedID             108 non-null    int64 \n",
      " 3   Title                108 non-null    object\n",
      " 4   Abstract             108 non-null    object\n",
      " 5   Quinone              108 non-null    object\n",
      " 6   Keyword Sentence     108 non-null    object\n",
      " 7   Species and Genus    108 non-null    object\n",
      "dtypes: int64(1), object(7)\n",
      "memory usage: 7.6+ KB\n"
     ]
    }
   ],
   "source": [
    "### Building Quinone Dictionary\n",
    "df = pd.read_csv(\"output50.csv\")\n",
    "df.info()\n",
    "df.head()\n",
    "df.tail()\n",
    "\n",
    "quinones = {\n",
    "    \"q\": [\"quinone\", \n",
    "          \"{},{}-benzoquinone\",\"{},{}-bq\", \n",
    "          \"{},{}-dimethyl-{},{}-benzoquinone\",\"{},{}-dimethyl-{},{}-bq\", \"{},{}-d-{},{}-benzoquinone\",\"{},{}-d-{},{}-bq\", \n",
    "          \"{}-methyl-{},{}-naphthoquinone\", \"{}-methyl-{},{}-nq\", \"{}-m-{},{}-naphthoquinone\", \"{}-m-{},{}-nq\",\n",
    "          \"{},{}-dimethoxy-{},{}-naphthoquinone\", \"{},{}-dimethoxy-{},{}-nq\",\"{},{}-d-{},{}-naphthoquinone\", \"{},{}-d-{},{}-nq\",\n",
    "          \"{},{},{},{}-tetramethyl-{},{}-benzoquinone\", \"{},{},{},{}-tetramethyl-{},{}-bq\", \"{},{},{},{}-t-{},{}-benzoquinone\", \"{},{},{},{}-t-{},{}-bq\",],\n",
    "    \"uq\": [\"ubiquinone\", \"coenzyme Q\", \"coQ\", \"UQ\", \"Ubiquinone{}\", \"Coenzyme Q{}\", \"CoQ{}\", \"uq{}\", \"coQ{}\", \"UQ{}\", \"Ubq{}\", \"CoQ{}\", \"CoQ{}\"],\n",
    "    \"uq-{}\": [\"UQ{}\", \"UQ{}H{}\", \"ubiquinone-{}\", \"ubiquinone-{}(H{})\", \"Ubiquinone-{}\", \"Ubiquinone-{}(H{})\", \"uq-{}\", \"uq{}H{}\", \"Ubq-{}\", \"Ubq{}(H{})\"],\n",
    "    \"mk\": [\"menaquinone\", \"vitamin K2\", \"MK\", \"Menaquinone\", \"Vitamin K2\", \"mk\", \"vK2\", \"MK\", \"Mk\", \"VK2\"],\n",
    "    \"mk-{}\": [\"MK{}\", \"MK{}H{}\", \"menaquinone-{}\", \"menaquinone-{}(H{})\", \"Menaquinone-{}\", \"Menaquinone-{}(H{})\", \"mk-{}\", \"mk{}H{}\", \"Mk-{}\", \"Mk{}(H{})\"],\n",
    "    \"pq\": [\"plastoquinone\", \"PQ\", \"Plastoquinone\", \"pq\", \"Pq\"],\n",
    "    \"rhq\": [\"rhodoquinone\", \"RQ\", \"Rhodoquinone\", \"rhq\", \"Rhq\"],\n",
    "    \"caldq\": [\"Caldariella quinone\", \"CQ\", \"Caldariella Quinone\", \"Caldariellaquinone\", \"caldq\", \"Caldq\", \"Cq\"],\n",
    "    \"dmk\": [\"demethyl-menaquinone\", \"DMK\", \"Demethyl-menaquinone\", \"dmk\", \"Dmk\"],\n",
    "    \"mmk\": [\"methyl-menaquinone\", \"MMK\", \"Methyl-menaquinone\", \"mmk\", \"Mmk\"],\n",
    "    \"{}-mmk\": [\"{}-methyl-menaquinone\", \"{}-MMK\", \"{}-methyl-Menaquinone\", \"{}-Methyl-menaquinone\", \"{}-mmk\", \"{}-Mmk\"],\n",
    "    \"{},{}-dmmk\": [\"{},{}-dimethyl-menaquinone\", \"{},{}-DMMK\", \"{},{}-Dimethyl-menaquinone\", \"{},{}-dmmk\", \"{},{}-Dmmk\"],\n",
    "}\n",
    "\n",
    "### Extracting Quninones, Species and Genus, and Keywords\n",
    "# regex pattern to match species and genus names\n",
    "species_genus_pattern = re.compile(r\"[A-Z][a-z]*\\s[a-z]*\")\n",
    "\n",
    "# function to extract species and genus names from a title\n",
    "def extract_species_genus(title):\n",
    "    return \", \".join(species_genus_pattern.findall(title))\n",
    "\n",
    "# function to find sentences containing a keyword in a given text\n",
    "def find_keyword_sentence(text, keyword):\n",
    "    sentences = sent_tokenize(text)\n",
    "    return \"; \".join(sentence for sentence in sentences if keyword.lower() in sentence.lower())\n",
    "\n",
    "# function to check if a row contains any quinone terms\n",
    "def contains_quinone(row):\n",
    "    matches = {}\n",
    "    keyword_sentences = set()\n",
    "\n",
    "    # iterate over quinone terms\n",
    "    for k, v in quinones.items():\n",
    "        for term in v:\n",
    "            # create a regex pattern for term\n",
    "            pattern = term.format(*[\"\\d+\"] * term.count(\"{}\"))\n",
    "            regex = re.compile(pattern, re.IGNORECASE)\n",
    "\n",
    "            # search for term in title and abstract of row\n",
    "            match_in_title = regex.search(row[\"Title\"])\n",
    "            match_in_abstract = regex.search(row[\"Abstract\"])\n",
    "\n",
    "            # if there is a match, store matched text and related information\n",
    "            if match_in_title or match_in_abstract:\n",
    "                match_text = match_in_title.group(0) if match_in_title else match_in_abstract.group(0)\n",
    "\n",
    "                if match_text not in matches:\n",
    "                    matches[match_text] = []\n",
    "                matches[match_text].append(k)\n",
    "\n",
    "                # find keyword sentences containing matched text in title and abstract\n",
    "                keyword_sentence_title = find_keyword_sentence(row[\"Title\"], match_text)\n",
    "                keyword_sentence_abstract = find_keyword_sentence(row[\"Abstract\"], match_text)\n",
    "\n",
    "                if keyword_sentence_title:\n",
    "                    keyword_sentences.add(keyword_sentence_title)\n",
    "                if keyword_sentence_abstract:\n",
    "                    keyword_sentences.add(keyword_sentence_abstract)\n",
    "\n",
    "    # if there are matches, return matches and related keyword sentences\n",
    "    if matches:\n",
    "        matches_str = \", \".join(f\"'{k}': ({', '.join(matches[k])})\" for k in matches)\n",
    "        return matches_str, \"; \".join(keyword_sentences)\n",
    "    else:\n",
    "        return None, None\n",
    "\n",
    "\n",
    "df = pd.read_csv(\"output50.csv\")\n",
    "df.fillna(\"\", inplace=True)\n",
    "\n",
    "# apply contains_quinone function to each row of DataFrame and expand results into columns\n",
    "df[[\"Quinone\", \"Keyword Sentence\"]] = df.apply(contains_quinone, axis=1, result_type=\"expand\")\n",
    "\n",
    "# extract species and genus names from titles\n",
    "df[\"Species and Genus\"] = df[\"Title\"].apply(extract_species_genus)\n",
    "\n",
    "# df['Journal Name'].describe()\n",
    "\n",
    "df = df[(~df[\"Quinone\"].isnull())] # df with just quinone matches\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'quinone': 178, 'menaquinone': 138, 'ubiquinone': 76, 'Menaquinone': 5})\n"
     ]
    }
   ],
   "source": [
    "\n",
    "## keyword analysis - create a counter dictionary and update it with quinones from data\n",
    "## taxonomy classification - NCBI's API for taxonony data?\n",
    "## quinone length analysis\n",
    "## visualization \n",
    "\n",
    "# statistics - on taxonomy distribution, chains of quinones, lengths varying different genus/order, variablility in the quinone chain.\n",
    "\n",
    "# crosscheck species/genus mentioned in title with abstract to find sentence where that specific species/genus was mentioned.\n",
    "### check what to do with pubmedResults.txt file\n",
    "# classify organisms using the taxonomy based on species and genus names extracted.\n",
    "# the hieacal classification order - missing species\n",
    "# look for artices give a species, validate quinone info.\n",
    "# get the tail length in a new column.\n",
    "\n",
    "# determine frequency of occurrence of each type of quinone in data set.\n",
    "from collections import Counter\n",
    "from nltk.tokenize import word_tokenize\n",
    "\n",
    "quinone_counter = Counter()\n",
    "\n",
    "for quinone_description in df[\"Quinone Description\"]:\n",
    "    tokens = word_tokenize(quinone_description)\n",
    "    quinone_counter.update([token for token in tokens if token in quinone_types])\n",
    "\n",
    "print(quinone_counter)\n",
    "\n",
    "# identify relationships between types of quinones found and species or genus of organisms.\n",
    "# look for patterns or trends in data that might indicate a relationship between length of quinone tail and species or genus of organism.\n",
    "# create visualizations to better understand and communicate these patterns or relationships.\n",
    "# based on these patterns or relationships, make predictions about types of quinones that might be found in other species or genus.\n",
    "\n",
    "# using 'Species and Genus' classify organisms according to their taxonomy \n",
    "# link organisms to specific quinones using 'Quinone Found' \n",
    "# understand quinone structure: information about type of quinone and length of quinone tail can be used to study structural diversity of quinones across different organisms.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# formating outputs\n",
    "outputFile = open(\"pubMedResults50.txt\", \"w+\")\n",
    "for count, (index, row) in enumerate(df.iterrows(), start=1):\n",
    "    outputFile.write(f\"{count}.Journal: {row['Journal Name']}\\n\\tPubMedID: {row['PubMedID']}\\n\\tSpecies and Genus: {row['Species and Genus']}\\n\\tQuinone Found: {row['Quinone']}\\n\\t{row['Keyword Sentence']}\\n\\n\")\n",
    "outputFile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Feedback on your intermediate report:\n",
    "\n",
    "# Please provide a title: \n",
    "    #Taxonomy Classification and Quinone Analysis from Scientific Publications: A Natural Language Processing Approach\n",
    "# The report is clear and fulfills the requirements, but it contains many typos that should be corrected.\n",
    "\n",
    "# There is no explicit related work section, and only one reference is given, this has to be strengthened.\n",
    "# State of the art - explain the importance of work and relate it to how the last real research done on this was from the 1981 paper\n",
    "# quinones are sometimes used as taxonomic markers - specific quiones can be link to specific species\n",
    "# molecule that is a diagnostic for certain species\n",
    "# look at relation same species order have same quinones\n",
    "\n",
    "# There is a figure that reports experiments but no details are given. - explain to novice reader\n",
    "# What was the motivation and the experimental conditions?\n",
    "# Unique topic, computer science to improve the field of biology\n",
    "# biological context - info, importantance, understand of quinone distrubution lilited. info in journals, microbiology intrest in quinones, TIMC motivations - pathways oninone biosynthese, resparatory chains, qunone type breath different substrates, what conditions organism can grow and colonize. quinone and organism give a hint of what conditions are used to grow them\n",
    "\n",
    "# team - charaterizing quinone pathways by identifying genes in production\n",
    "# confront with team data - validate, discover new pathways? \n",
    "# figures to explain the pathways, variants, etc.\n",
    "# very expiremental work - plenty of surprises\n",
    "\n",
    "# The purpose is to extract all this uncollected quinone data and compare it to research, in a way validating the journals and updating any new discoveries or information previously unknown.\n",
    "\n",
    "# species -> quinones\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
